Source/Core/Core/HW/DSP.cpp

```diff
@@ -472,140 +490,138 @@ void UpdateAudioDMA()
-        {
-          Memory::Write_U64_Swap(*(u64*)&s_ARAM.ptr[s_arDMA.ARAddr & s_ARAM.mask], s_arDMA.MMAddr);
-        }
-        else if ((s_ARAM_Info.Hex & 0xf) == 4)
-        {
-          Memory::Write_U64_Swap(*(u64*)&s_ARAM.ptr[s_arDMA.ARAddr & s_ARAM.mask], s_arDMA.MMAddr);
-        }
-        else
-        {
-          Memory::Write_U64_Swap(*(u64*)&s_ARAM.ptr[s_arDMA.ARAddr & s_ARAM.mask], s_arDMA.MMAddr);
-        }
+template <u32 MB_SIZE>
+static std::optional<u32> ARAM_map_to_16MB(u32 address)
+{
+  static_assert((MB_SIZE == 2) || (MB_SIZE == 4) || (MB_SIZE == 8) || (MB_SIZE == 16) ||
+                (MB_SIZE == 32));
 
-        s_arDMA.MMAddr += 8;
-        s_arDMA.ARAddr += 8;
-        s_arDMA.Cnt.count -= 8;
-      }
-    }
-    else
-    {
-      // Assuming no external ARAM installed; returns zeros on out of bounds reads (verified on real
-      // HW)
-      while (s_arDMA.Cnt.count)
-      {
-        Memory::Write_U64(0, s_arDMA.MMAddr);
-        s_arDMA.MMAddr += 8;
-        s_arDMA.ARAddr += 8;
-        s_arDMA.Cnt.count -= 8;
-      }
-    }
+  address &= ARAM_MEMORY_WRAP_MASK;
+  if (address >= MB_SIZE * 1024 * 1024)
+  {
+    return std::nullopt;
+  }
+  if constexpr (MB_SIZE < 16)
+  {
+    return ((address & ARAM_UPPER_MASK1) << 1) | (address & ARAM_LOWER_MASK1);
+  }
+  else if constexpr (MB_SIZE > 16)
+  {
+    return ((address & ARAM_UPPER_MASK2) >> 1) | (address & ARAM_LOWER_MASK2);
   }
   else
   {
-    // MRAM -> ARAM
-    DEBUG_LOG(DSPINTERFACE, "DMA %08x bytes from MRAM %08x to ARAM %08x PC: %08x",
-              s_arDMA.Cnt.count, s_arDMA.MMAddr, s_arDMA.ARAddr, PC);
+    return address;
+  }
+}
+
+using ARAM_ADDRESS_CONVERSION_F = std::optional<u32> (*)(u32 address);
+constexpr ARAM_ADDRESS_CONVERSION_F conversion_functions[8] = {
+    &ARAM_map_to_16MB<2>,  &ARAM_map_to_16MB<4>,  &ARAM_map_to_16MB<8>,  &ARAM_map_to_16MB<16>,
+    &ARAM_map_to_16MB<32>, &ARAM_map_to_16MB<32>, &ARAM_map_to_16MB<32>, &ARAM_map_to_16MB<32>,
+};
+
+enum
+{
+  ARAM_DMA_DIR_TO_ARAM = 0,
+  ARAM_DMA_DIR_FROM_ARAM = 1,
+};
 
-    // Incoming data into ARAM is mirrored every 64MB (verified on real HW)
-    s_arDMA.ARAddr &= 0x3ffffff;
-    s_arDMA.MMAddr &= 0x3ffffff;
+// Size of the smallest unit of transfer to/from ARAM via DMA.
+constexpr u32 ARAM_LINE_SIZE = 0x20;
+
+// Maximum number of lines to transfer at a time via DMA.
+constexpr u32 ARAM_MAX_TRANSFER_CHUNKING = 0x10;
+
+// The number of clock ticks for each line to be transferred.
+constexpr u32 TICKS_TO_TRANSFER_LINE = 246;
+
+static void Do_ARAM_DMA()
+{
+  constexpr std::array<const char*, 2> aram_transfer_direction = {"to", "from"};
 
-    if (s_arDMA.ARAddr < s_ARAM.size)
+  s_dspState.DMAState = 1;
+  // ARAM is mirrored every 64MB (verified on real HW) - done in address conversion func
+  // Source/destination/count aligned to 32 bytes      - done in MMIO handler
+
+  u32 lines_to_transfer = std::min(s_arDMA.Cnt.count / ARAM_LINE_SIZE, ARAM_MAX_TRANSFER_CHUNKING);
```

Comment:
Nit: Please keep the naming style consistent.

Reply:
was this mostly about ticksToTransfer not being snake case? (fixed)